### PR TITLE
create initial connection

### PR DESCRIPTION
**Pull request by sergiyv-bitquill**
### Summary

MySQL - JDBC - Optimization - Extend Plugin Interface for Initial Connection Method

### Additional Reviewers

@karenc-bq 
@ColinKYuen

### CODE REVIEW

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 1]
```diff
@@ -263,11 +265,15 @@ private Object executeMethodDirectly(String methodName, Object[] args) {
...
   }
 
-  protected void initProxy(Function<Log, ConnectionPluginManager> connectionPluginManagerInitializer)
+  protected void initPluginManager(Function<Log, ConnectionPluginManager> connectionPluginManagerInitializer, ConnectionUrl connectionUrl)
       throws SQLException {
     if (this.pluginManager == null) {
       this.pluginManager = connectionPluginManagerInitializer.apply(log);
       this.pluginManager.init(this, connProps);
+
+      if(this.currentConnection == null) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:274]:
```suggestion
      if (this.currentConnection == null) {
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/DefaultConnectionPlugin.java [hunk 2]
```diff
@@ -71,6 +89,17 @@ public Object execute(
     }
   }
 
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    if(this.currentConnectionProvider.getCurrentConnection() != null) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/DefaultConnectionPlugin.java:94]:
```suggestion
    if (this.currentConnectionProvider.getCurrentConnection() != null) {
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 3]
```diff
@@ -180,16 +181,15 @@ public FailoverConnectionPlugin(
...
-
     if (!this.enableFailoverSetting) {
       return;
     }
 
-    AuroraTopologyService topologyService = topologyServiceSupplier.get();
-    topologyService.setPerformanceMetricsEnabled(this.gatherPerfMetricsSetting);
+    this.topologyService = topologyServiceSupplier.get();
+    if(this.topologyService instanceof  ICanCollectPerformanceMetrics) {
+      ((ICanCollectPerformanceMetrics)topologyService).setPerformanceMetricsEnabled(this.gatherPerfMetricsSetting);
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:190]:
```suggestion
    if (this.topologyService instanceof ICanCollectPerformanceMetrics) {
      ((ICanCollectPerformanceMetrics) topologyService).setPerformanceMetricsEnabled(this.gatherPerfMetricsSetting);
```
Are we checking if topology is instance of `ICanCollectPerformanceMetrics` for future proofing if someone wants to create their own topology class?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 4]
```diff
@@ -956,6 +1059,38 @@ private void initTopology() {
     }
   }
 
+  private void createConnection(ConnectionUrl connectionUrl) throws SQLException {
+
+    // Connection isn't created - try to use cached topology to create it
+    if(this.currentConnectionProvider.getCurrentConnection() == null) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:1065]:
```suggestion
    if (this.currentConnectionProvider.getCurrentConnection() == null) {
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 5]
```diff
@@ -956,6 +1059,38 @@ private void initTopology() {
...
+        try {
+          attemptConnectionUsingCachedTopology();
+        } catch (SQLException e) {
+          // do nothing - attempt to connect directly will be made below
+        }
+      }
+    }
+
+    // Connection isn't created - let other plugins to create it
+    if(this.currentConnectionProvider.getCurrentConnection() == null) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:1083]:
```suggestion
    if (this.currentConnectionProvider.getCurrentConnection() == null) {
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 6]
```diff
@@ -956,6 +1059,38 @@ private void initTopology() {
...
+      this.nextPlugin.openInitialConnection(connectionUrl);
+    }
+
+    // Connection isn't created - take it over and create it
+    if(this.currentConnectionProvider.getCurrentConnection() == null) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:1088]:
```suggestion
    if (this.currentConnectionProvider.getCurrentConnection() == null) {
```

**Review by ColinKYuen — APPROVED** [review]
Just nits

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java (hunk 3 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java]: Implementing ICanCollectPerformanceMetrics is optional. If topology service implements it then the driver can engage it.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 7]
```diff
@@ -182,7 +187,7 @@ public synchronized Object invoke(Object proxy, Method method, Object[] args)
           this.currentConnection.getClass(),
           methodName,
           () -> method.invoke(currentConnection, args),
-          args);
+          Arrays.copyOf(args, args.length));
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:190]: I don't think we need to explicitly call `Arrays.copyOf` here. Modifying args shouldn't affect the args used in `method.invoke`

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 8]
```diff
@@ -110,19 +113,21 @@ public ConnectionProxy(ConnectionUrl connectionUrl, JdbcConnection connection) t
   public static JdbcConnection autodetectClusterAndCreateProxyInstance(ConnectionUrl connectionUrl)
       throws SQLException {
 
-    final IConnectionProvider connectionProvider = new BasicConnectionProvider();
-    final ConnectionProxy connProxy = new ConnectionProxy(connectionUrl, connectionProvider.connect(connectionUrl.getMainHost()));
-    if (connProxy.isPluginEnabled()) {
+    boolean pluginsEnabled = Boolean.parseBoolean(
+      connectionUrl.getConnectionArgumentsAsProperties()
+        .getProperty(PropertyKey.useConnectionPlugins.getKeyName(), Boolean.toString(true)));
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:118]:
```suggestion
    boolean pluginsEnabled = this.connProps.getBooleanProperty(PropertyKey.useConnectionPlugins).getValue();
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/IConnectionPlugin.java [hunk 9]
```diff
@@ -39,6 +42,6 @@ Object execute(
       Callable<?> executeSqlFunc,
       Object[] args)
       throws Exception;
-
+  void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException;
   void releaseResources();
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/IConnectionPlugin.java:46]:
```suggestion
  void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException;
  
  void releaseResources();
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 10]
```diff
@@ -177,12 +182,14 @@ public synchronized Object invoke(Object proxy, Method method, Object[] args)
       return executeMethodDirectly(methodName, args);
     }
 
+    Object[] argsCopy = args == null ?  null : Arrays.copyOf(args, args.length);
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:185]: Don't think we need this

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java [hunk 11]
```diff
@@ -180,16 +181,15 @@ public FailoverConnectionPlugin(
...
 
-    this.connectionProvider = connectionProvider;
-
     if (!this.enableFailoverSetting) {
       return;
     }
 
-    AuroraTopologyService topologyService = topologyServiceSupplier.get();
-    topologyService.setPerformanceMetricsEnabled(this.gatherPerfMetricsSetting);
+    this.topologyService = topologyServiceSupplier.get();
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java:188]: The supplier signature needs to be changed to `Supplier<ITopologyService> topologyServiceSupplier`

**Comment by karenc-bq** [comment]
Can you please enable tests `testForWriterReconnectWhenDirectReaderConnectionFails` and `testForWriterReconnectWhenInvalidInitialWriterConnection` in `FailoverConnectionPluginTest?

For testForWriterReconnectWhenInvalidInitialWriterConnection, you will have to add the following lines after `initFailoverPlugin` (L271):
```java
    final ConnectionUrl connectionUrl =
        ConnectionUrl.getConnectionUrlInstance(url, new Properties());
    failoverPlugin.openInitialConnection(connectionUrl);
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java (hunk 8 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:118]: It's not possible to use "this" in static method.

**Review by karenc-bq — APPROVED** [review]
Can you please final the variables before you merge

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java (hunk 10 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:185]: As discussed. Let's keep it to avoid side effects.

**Pull request merged**